### PR TITLE
WIP: new DFT api

### PR DESCRIPTION
**Pull request by stevengj**
This is not quite ready to merge yet, but I wanted to bring it to your attention to see what you think.  It is a rewritten DFT API that provides:
- `p = plan_fft(x)` (and similar) now returns a subtype of `Base.DFT.Plan`.  It acts like a linear operator: you can apply it to an existing array with `p * x`, and apply it to a preallocated output array with `A_mul_B!(y, p, x)`.  You can apply the inverse plan with `p \ x` or `inv(p) * x`.
- It is easy to add new FFT algorithm implementations for specific numeric and array types...

Partly since we have discussed moving FFTW to a module at some point for licensing reasons, and partly to support numeric types like `BigFloat`, the patch also includes a pure-Julia FFT implementation, with the following features.
- One-dimensional transforms of arbitrary `AbstractVector`s of arbitrary `Complex` types (including `BigFloat`), with fast algorithms for all sizes (including prime sizes).  Multidimensional transforms of arbitrary `StridedArray` subtypes.
- Uses code generation (a baby version of the FFTW code generator) in Julia to generate kernels for use as base cases of the transforms; algorithms in general are FFTW1-like.
- Performance is reasonable. For moderate-power-of-two sizes, it is within a factor of 2 or so of FFTW, with the difference mainly due to the lack of SIMD.   ~~For non-power-of-two sizes, the performance is several times worse, but I suspect that there may be some fixable compiler problem there (a good stress test for the compiler).~~ ~~(Non-power-of-two sizes seem fine with latest Julia).~~ (Now they suck again.)
- `import FFTW` is enough to automatically switch over to the FFTW algorithms for supported types.

Major to-do items:
- [x] Documentation
- [x] More test cases (implemented a self-test algorithm by Funda Ergün)

Minor to-do items (will eventually be needed to move the FFTW code out to an external module, but should not block merge):
- [ ] Native Julia `rfft` support.
- [ ] Native Julia in-place transforms. (The 1d FFT algorithm here is out-of-place, but it could act "in-place" via a buffer.  The multi-dim FFTs only need a buffer of length `maximum(size(x))`.)
- [ ] Performance improvements in non-power-of-two transforms — these underwent a major performance regression at some point (possibly due to the inlining heuristics changing); the right thing is ultimately to change the generator to spit out kernels in terms of real rather than complex arithmetic (or just to use FFTW's generator).

cc: @JeffBezanson, @timholy

### CODE REVIEW

**Comment by stevengj** [comment]
Rebased.

**Comment by stevengj** [comment]
Hmm, I'm seeing some further type-inference problems, analogous to those in #9772.  Julia is having a hard time figuring out the FFT return types for some reason.

**Comment by tknopp** [comment]
+1 to merging. At the point where "Default packages" might become reality one will have to think if the fft code could be moved to a default package. But this is a later discussion.

**Comment by stevengj** [comment]
This is weird.  I'm getting
```
@inferred plan_fft(rand(Complex128,10), (1,))
ERROR: return type Base.DFT.FFTW.cFFTWPlan{Complex{Float64},-1,false} does not match inferred return type Base.DFT.FFTW.cFFTWPlan{Complex{Float64},K,false}
 in error at error.jl:19
```
even though `plan_fft` has an explicit type-assertion:
```
for (f,direction) in ((:fft,:FORWARD), (:bfft,:BACKWARD))
    plan_f = symbol(string("plan_",f))
    plan_f! = symbol(string("plan_",f,"!"))
    idirection = direction == :FORWARD ? :BACKWARD : :FORWARD
    @eval begin
        function $plan_f{T<:fftwComplex}(X::StridedArray{T}, region;
                                         flags::Integer=ESTIMATE,
                                         timelimit::Real=NO_TIMELIMIT)
            cFFTWPlan(X, fakesimilar(flags, X, T), region,
                      $direction, flags,timelimit)::cFFTWPlan{T,$direction,false}
        end
        ...
end
```
@simonster, any idea of why Julia would be disbelieving my type assertion here?

**Comment by simonster** [comment]
@stevengj As it seems you've figured out, `$direction` is being interpolated into the AST as an identifier, not a value, and we don't do constant folding during inference. It seems like you could use:

```julia
for (f,direction) in ((:fft,FORWARD), (:bfft,BACKWARD))
    plan_f = symbol(string("plan_",f))
    plan_f! = symbol(string("plan_",f,"!"))
    idirection = direction == FORWARD ? BACKWARD : FORWARD
    @eval begin
        function $plan_f{T<:fftwComplex}(X::StridedArray{T}, region;
                                         flags::Integer=ESTIMATE,
                                         timelimit::Real=NO_TIMELIMIT)
            cFFTWPlan(X, fakesimilar(flags, X, T), region,
                      $direction, flags,timelimit)::cFFTWPlan{T,$direction,false}
        end
        ...
end
```

**Comment by simonster** [comment]
Alternatively, I think `FORWARD` and `BACKWARD` could be types rather than constants, in which case you may not even need the typeassert.

**Comment by stevengj** [comment]
@simonster, the typeassert is still needed for other reason; in general, Julia's inference gets confused by how the `FFTWPlan` constructor decides on the type parameters.

**Comment by stevengj** [comment]
Another weird inference problem:
```
julia> @inferred Base.DFT.normalization(Float64, (10,), (1,))
ERROR: return type Float64 does not match inferred return type Any
 in error at error.jl:19

julia> @code_typed Base.DFT.normalization(Float64, (10,), (1,))
1-element Array{Any,1}:
 :($(Expr(:lambda, Any[:T,:sz,:region], Any[Any[:_var0,:_var1],Any[Any[:T,Type{Float64},0],Any[:sz,(Int64,),0],Any[:region,(Int64,),0],Any[:_var0,Array{Int64,1},18],Any[:_var1,Int64,18]],Any[]], :(begin  # dft.jl, line 106:
        _var0 = getindex(vcat((top(tupleref))(sz::(Int64,),1)::Int64)::Array{Int64,1},vcat((top(tupleref))(region::(Int64,),1)::Int64)::Array{Int64,1})::Array{Int64,1}
        _var1 = (top(_mapreduce))($(Expr(:new, :((top(getfield))(Base,:IdFun)::Type{IdFun}))),$(Expr(:new, :((top(getfield))(Base,:MulFun)::Type{MulFun}))),_var0::Array{Int64,1})::Int64
        return (top(box))(Float64,(top(div_float))((top(box))(Float64,(top(sitofp))(Float64,1)),(top(box))(Float64,(top(sitofp))(Float64,_var1::Int64))))
    end::Float64))))
```
The `@code_typed` output seems to indicate that it knows the output is `Float64`, but `@inferred` is complaining that it infers `Any`.

**Comment by simonster** [comment]
@stevengj That one is my fault. `@inferred` just calls `typeof` on the arguments, but for `Float64` that yields `DataType` rather than `Type{Float64}`. Type inference should work correctly there.

**Comment by stevengj** [comment]
No, I don't think it's just the `@inferred` macro:
```
julia> import Base.DFT: ScaledPlan, normalization
         foo(x, region) = ScaledPlan(plan_bfft(x, region), normalization(x, region))
         bar(x, region) = ScaledPlan(plan_bfft(x, region), 1.0)

julia> @inferred foo(z, 1)
ERROR: return type Base.DFT.ScaledPlan{Complex{Float64},Base.DFT.FFTW.cFFTWPlan{Complex{Float64},1,false},Float64} does not match inferred return type Base.DFT.ScaledPlan{T,P,N}
 in error at error.jl:19

julia> @inferred bar(z, 1)
1.0 * FFTW backward plan for 10-element array of Complex{Float64}
(dft-direct-10 "n2bv_10_sse2")
```
It looks like my `normalization` function really is defeating type inference somehow:
```
normalization(T, sz, region) = one(T) / prod([sz...][[region...]])
normalization(X, region) = normalization(real(eltype(X)), size(X), region)
```

**Comment by simonster** [comment]
There was a bug in `@inferred` (fixed in 393f2691fed45d2eb0bf2241404a92b3bd7be760), which fixes the issue in your previous comment, but there's still a type inference issue related to splatting the region when it is an Int:

```julia
julia> f() = [1...];

julia> code_typed(f, ())
1-element Array{Any,1}:
 :($(Expr(:lambda, Any[], Any[Any[],Any[],Any[]], :(begin  # none, line 1:
        return (top(_apply))(call,vcat,1)::Any
    end::Any))))
```

**Comment by ViralBShah** [comment]
@stevengj Is this likely to land for 0.4? Should we fix the `rfft` type stability issue (#9772) separately on master otherwise?

**Comment by tkelman** [comment]
This hasn't seen much movement recently so I'm going to guess it's out of scope for 0.4.

Thinking out loud, but does this need to be developed as a part of base? Are there separate parts of base that would need to tie into this implementation, or could it be done as a package to start with?

**Comment by stevengj** [comment]
If we want FFTs in base, then some form of this needs to be in base.

**Comment by tkelman** [comment]
That's the core question, do they need to be? They are now, but if there aren't major dependencies of other parts of base on FFT's, array operations and whatnot, then they don't need to be a dependency of all code written in the language that doesn't do FFT's. Granted the cost of a pure-Julia version is lower than that of the entire FFTW library, it's a relatively small amount of build time and system image size, but the constraints on development flexibility are probably the most significant here.

As a package, people hypothetically could have been using this for the past year for FFT's on arbitrary types, right? As a not-yet-merged PR to base it gets difficult to use every time it needs a rebase.

**Comment by ScottPJones** [comment]
Well, I'd say you really don't want anything that is GPL'ed (not LGPL, which should be OK to call) in base julia, ever... that can severely limit the viability of the language in any commercial environment...
FFTW is GPL (or at least, the version currently in Base), so by that criteria, I think it should go (before 0.4, IMO)

**Comment by tkelman** [comment]
> (before 0.4, IMO)

We have an optional `USE_GPL_LIBS` flag. 0.4 is too soon and removing large chunks of code from base is not going to happen before then.

What I'm actually thinking about is for the new pure-Julia version here (which is slower than FFTW on types that FFTW can support, but is being developed from scratch with an MIT license) could be put into a package now, which would let people get their hands on it and start using it before this PR gets merged.

It's the question of #5155 - is the Julia standard library trying to be a the minimum required for a general-purpose programming language, or also including all of a technical computing environment a la scipy or matlab or octave too. Right now it's the latter, but it doesn't always have to stay that way, especially if we (hopefully soon) get precompiled packages and (a little further off) come up with a good way of bundling "default packages" into the technical-computing-environment installer that most users will start with.

The cost of having this MIT-licensed pure-Julia FFT implementation as part of base is low, so I don't feel too strongly about this, but wanted to put the idea out there.

**Comment by tknopp** [comment]
@ScottPJones: I have stated it at another thread but making julia GPL free is absolutely trivial by simply removing the shared libraries. And if you want to distribute julia in a commercial environment you will use your own installer. Thus, it is not relevant what license "stock" julia is distributed.

> As a package, people hypothetically could have been using this for the past year for FFT's on arbitrary types, right? As a not-yet-merged PR to base it gets difficult to use every time it needs a rebase.

Thanks @tkelman this is exactly the important point that I also tried to make clear in our SparseVector discussion. I know, the case is a little different and don't want to processed the discussion here, I just want to highlight your sentence and its importance for any future discussion on base inclusion (and also #5155)

**Comment by ScottPJones** [comment]
>into the technical-computing-environment installer that most users will start with.

I know I'm still the outlier here... but I do think Julia has the potential to be so wonderful for C/C++/Java programmers or even scripting programmers (JS, Python) (it's already wonderful, just some rough edges), that I think that statement will not be true for much longer...  Maybe I'm just being overly optimistic... but judging from the reactions of the other programmers here... I don't think so...

**Comment by ScottPJones** [comment]
@tknopp  I think the problem is, that if people get the language, write an app using some features that are part of the base language, or packages that depend on those features, and *then* find out that it can't be distributed, because there is no non-GPL alternative, *that's* bad.
I'm now building with USE_GPL_LIB = 0, to give to the other developers, so that they don't accidentally depend on something that could bite us later on...  The other issue is the load time and resources that Julia takes up, with so much stuff in base, when otherwise you might be able to better use Julia as a great scripting language....

**Comment by mbauman** [comment]
Instead of wholesale removing the code, could we simply deprecate it in favor of an external package?  I don't have strong opinions here, but perhaps this could be a more sane code-removal path.  For example: deprecate FFTW in base for 0.4 in favor of the package, and then remove it for 0.5.

**Comment by StefanKarpinski** [comment]
@ScottPJones, nobody is against moving towards a model where the basic language is a general language without too much that is specific to numerical work, while having a standard numerical environment that includes additional features like FFTs – that's the whole point of #5155. However, ripping large parts of Base while we're trying to get to feature freeze is counterproductive. You can already build a Julia that is GPL-free, so the license issue is not convincing as a show-stopper.

**Comment by tkelman** [comment]
Crap, sorry, that might be a record for acceleration of a PR from stopped to derailed. Continued discussion should be in #5155.

**Comment by ScottPJones** [comment]
@mbauman, sounds good to me... the deprecation warning would solve the issue of accidentally depending on something that is GPLed... if this approach will work.
@StefanKarpinski I didn't say it was a show-stopper, just that, *IMO* it would be good to be able to remove it (better deprecated) before 0.4 release... (and yes, it is pretty close to release, wouldn't want to hold that up, at least after static precompilation is done)

**Comment by tknopp** [comment]
@ScottPJones: Really, for someone how wants to use Julia in a commercial environment it is not a big deal to check which components are GPL an which not. And to make this concrete: I work with a company where we discuss integrating code I have written in Julia into a commercial software. In these discussions it was never a big deal that stock Julia has GPL components. 

Back to topic: I also think it would be good to move things into two separate packages. One generic fft package and one fftw  package.

**Comment by tkelman** [comment]
> For example: deprecate FFTW in base for 0.4 in favor of the package, and then remove it for 0.5.

@stevengj can say better than I can, but that's a much more aggressive timetable than I was imagining. Since there has been little movement here I'm under the assumption that the FFT status quo will remain as it is for 0.4, unless this PR gets into mergeable condition in the next few weeks.

**Comment by tknopp** [comment]
I would also say that there don't need to be an action in 0.4.

**Comment by mbauman** [comment]
Yeah, I actually first wrote 0.5/0.6, but since we don't actually have a 0.6 milestone yet I figured I'd use the next two releases as an example.  Like I said, I don't have strong feelings, but I think deprecation could be much smoother than some of our other attempts.

**Comment by StefanKarpinski** [comment]
No, we're not fussing with FFTW right now. Could potentially happen in Arraymageddon.

**Comment by ScottPJones** [comment]
@StefanKarpinski Do you see 0.5 and Arraymageddon lasting very long?  A full year?  Until next summer's JuliaCon?  I'm fine with this not happening in 0.4, as long people agree that the direction in the reasonably near future is a fast loading general purpose Julia, with a standard "numerical" Julia, and precompiled modules that can bring general purpose Julia up to "numerical" Julia without any fuss.
(and I'll do whatever I can to help out).
@tknopp I know, but I still have to build now a GPL free Julia for the other developers, instead of them just downloading the standard version... not really a biggie (many thanks to @ViralBShah for adding USE_GPL_LIB)

**Comment by tknopp** [comment]
I don't know who "the other developers" are but its absolutely no problem to use GPL libraries in-house in a company. The first time the question is put on a table is when you start deploying that stuff. And if you distribute you own Julia program you will likely put the things into a custom system image in order to hide the source code.

Your request is absolutely fine. I just want to argue that I do not see a pressing need currently as a company distributing Julia will compile it anyway on its own. So in short: Nothing is stopping anybody from julia used in commercial programs right now.

**Comment by ScottPJones** [comment]
@tknopp I didn't say that it was *stopping*, just that it makes it harder, and it would be nice to try to lower *any* barriers to adoption of Julia (speed of loading, for scripts or lots of background jobs [this is an issue for us already], debugging, memory usage, performance in non "numeric/technical" areas... lots of things)
The nice thing I guess is, there's a *lot* of interesting work left to be done in Julia :grinning:

**Comment by ScottPJones** [comment]
OT: @tknopp The other developers are the people in the startup I'm consulting for here in Belgium...
Giving the way things are going, we may be looking for people with Julia expertise as the startup grows (maybe we'll be fighting Julia Computing for the best and brightest!)

**Comment by simonbyrne** [comment]
+1 for a package, even if only temporary.

**Comment by simonster** [comment]
I would actually like to get this into 0.4 if it's still possible. @stevengj Besides fixing the type inference problem above and rebasing (which I'm willing to take a stab at if you're busy), what else is left here? FWIW, I think that:

```julia
normalization(T, sz, region) = one(T) / prod(collect(sz)[collect(region)])
normalization(X, region) = normalization(real(eltype(X)), size(X), region)
```

should be inferred correctly.

**Comment by stevengj** [comment]
@simonster, I just rebased and am compiling now; will push the rebased version once tests pass.

**Comment by stevengj** [comment]
Okay, I've rebased, but the `@inferred` tests at the end of `test/fft.jl` are failing because of a type-inference problem (`ERROR: LoadError: LoadError: return type Array{Complex{Float64},1} does not match inferred return type Any`).  @simonster, would you mind taking a look?

File: test/fft.jl [hunk 1]
```diff
@@ -64,153 +27,299 @@ true_fftd2_m4 = [
...
+true_fftd3_m3d = Array(Float32, 5, 3, 2)
+true_fftd3_m3d[:,:,1] = 17:2:45
+true_fftd3_m3d[:,:,2] = -15
+
+# use invoke to force usage of CTPlan versions even if FFTW is present
+for A in (Array,SubArray)
+    for f in (:fft,:ifft,:plan_fft,:plan_ifft)
+        f_ = symbol(string(f, "_"))
+        @eval begin
+            $f_{T,N}(x::$A{T,N}) = invoke($f, (AbstractArray{T,N},), x)
```
Comment (simonster) [inline test/fft.jl]: This should be `invoke($f, Tuple{AbstractArray{T,N}}, x)` and similarly below. With this change the inference tests pass for me, although there are a few deprecation warnings that need to be fixed.

**Comment by ScottPJones** [comment]
What about the checkboxes in the opening comment? Does that need to be updated? Seems like a good step

File: base/utf8proc.jl [hunk 2]
```diff
@@ -75,10 +75,7 @@ function utf8proc_map(s::ByteString, flags::Integer)
                    s, sizeof(s), p, flags)
     result < 0 && error(bytestring(ccall(:utf8proc_errmsg, Ptr{UInt8},
                                          (Cssize_t,), result)))
-    a = ccall(:jl_ptr_to_array_1d, Vector{UInt8},
-              (Any, Ptr{UInt8}, Csize_t, Cint),
-              Vector{UInt8}, p[], result, true)
-    ccall(:jl_array_to_string, Any, (Any,), a)::ByteString
+    pointer_to_string(p[], result, true)::ByteString
```
Comment (tkelman) [inline base/utf8proc.jl]: unrelated change?

File: base/dft.jl [hunk 3]
```diff
@@ -0,0 +1,199 @@
+module DFT
```
Comment (tkelman) [inline base/dft.jl]: license headers on new files

File: test/fft.jl (hunk 1 above)
Comment (stevengj) [inline test/fft.jl]: Ah, thanks.  Though it seems weird that it works but gives a type-inference failure.

File: base/utf8proc.jl (hunk 2 above)
Comment (stevengj) [inline base/utf8proc.jl]: No, it's not unrelated.  What happened is that I needed exactly the same code in FFTW.jl, and rather than copy-and-paste the code it seemed better to put it into a function, analogous to `pointer_to_array`.

File: base/dft.jl (hunk 3 above)
Comment (stevengj) [inline base/dft.jl]: fixed

**Comment by ScottPJones** [comment]
I really like the way this allows the GPL FFTW stuff to be brought in later, and the code has more comments than I've seen elsewhere in Base.  Looking at the code (*really* over a 1000 new lines, no exaggeration!)  It looks to me like the 3rd (rfft in Julia) and 5th (performance of non-power-of-2 xforms) have been addressed, shouldn't they be checked off?  I can't tell about the 4th.
I'm sure this was done so long ago give the low PR number, that Docile may not have been available, and there are issues with `@doc` in `Base`, but could the methods and types possibly get doc strings?
Maybe a new macro could be added (for now), `@basedoc`, which would simply accept the `@doc` syntax, but just throw away the arguments, until we get `@doc` working as has been discussed elsewhere.
Are there any gists for the code for the various benchmarks that have been run, so that they can be re-run, as things change in this PR, and also in Base?   It would really be nice to see some good micro benchmarking of this, and I'm esp. interested in seeing how you produced the nice graph comparing different micro benchmark results... (I did see the thing about running some FFTW `bench` program, but nothing about the graphing... maybe that seems trivial to all of you!)
Finally... is there some rule among Julia developers that most variable names must be one or two characters?  It would be nice if somewhat longer ones (that generally aren't substrings of other variable names, at least in the same method) were used... I like to ^S (emacs incremental search) to find occurrences of a variable in a function, all the single character names make that impossible...
Other than the above, LGTM, and I hope this gets merged soon.

**Comment by carnaval** [comment]
(I like short variable names. you can use `C-M-s \<x\>` to look for x as an identifier)

**Comment by ScottPJones** [comment]
Well, that optimizes for the time spent *writing* the code, not the time spent *maintaining* the code...
`^Sval is faster and easier than C-M-s \<x\>, and somewhat longer names are also easier when eyeballing the code.   If you have successful software, you spend *many* more years maintaining code than you do writing it...

**Comment by hayd** [comment]
@ScottPJones OT I don't think the documentation situation is fully resolved yet.... atm ?help will work and online documentation only works from the RSTs (I think). I have some code which converts all the RSTs to markdown julia docstrings... but we need a way to get the online documentation looking up that (otherwise we'd be duplicating documentation). :(

**Comment by timholy** [comment]
@ScottPJones, I bet much of the habit of short variable names comes from the mathematics behind many of Julia's functions. Especially because unicode is allowed, the code often uses the same variable names as the publication that describes the algorithm/mathematics---and in math, all variables are short.

And then once you get used to that...

**Comment by ScottPJones** [comment]
@timholy I don't mind when it is a single character mathematical symbol (as long as they aren't things like 𝟙 :grinning:), because then it's still easy to search for with the editor, or by eyeball (if it even shows up as anything but a box!).
I still think it is poor programming practice, and most every style guide I've seen only allows some exceptions (in C/C++/Java guides) for i,j,k,l,m,n when the scope is very local...
I've found 3 characters pretty much a good minimum... along with not allowing variable names that are substrings of another name in the same function...

**Comment by stevengj** [comment]
@ScottPJones, it has the framework for interfacing a pure-Julia rfft or in-place fft, but there is no backend implementation (other than the one provided by FFTW) yet.

File: base/fft/FFTW.jl [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+const REDFT10 = 5
+const REDFT11 = 6
+const RODFT00 = 7
+const RODFT01 = 8
+const RODFT10 = 9
+const RODFT11 = 10
+
+let k2s = Dict{Int,ASCIIString}(R2HC => "R2HC", HC2R => "HC2R", DHT => "DHT", REDFT00 => "REDFT00", REDFT01 => "REDFT01", REDFT10 => "REDFT10", REDFT11 => "REDFT11", RODFT00 => "RODFT00", RODFT01 => "RODFT01", RODFT10 => "RODFT10", RODFT11 => "RODFT11")
+    global kind2string
+    kind2string(k::Integer) = k2s[int(k)]
```
Comment (simonster) [inline base/fft/FFTW.jl]: `int` is deprecated

File: base/fft/FFTW.jl [hunk 5]
```diff
@@ -0,0 +1,781 @@
...
+                     PlanPtr,
+                     (Int32, Ptr{Int}, Int32, Ptr{Int},
+                      Ptr{$Tr}, Ptr{$Tr}, Ptr{Int32}, UInt32),
+                     size(dims,2), dims, size(howmany,2), howmany,
+                     X, Y, knd, flags)
+        set_timelimit($Tr, NO_TIMELIMIT)
+        if plan == C_NULL
+            error("FFTW could not create plan") # shouldn't normally happen
+        end
+        return r2rFFTWPlan(plan, flags, R, X, Y, tuple(int(knd)...))
```
Comment (simonster) [inline base/fft/FFTW.jl]: Also deprecated

File: base/fft/FFTW.jl [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+                     PlanPtr,
+                     (Int32, Ptr{Int}, Int32, Ptr{Int},
+                      Ptr{$Tc}, Ptr{$Tc}, Ptr{Int32}, UInt32),
+                     size(dims,2), dims, size(howmany,2), howmany,
+                     X, Y, knd, flags)
+        set_timelimit($Tr, NO_TIMELIMIT)
+        if plan == C_NULL
+            error("FFTW could not create plan") # shouldn't normally happen
+        end
+        return r2rFFTWPlan(plan, flags, R, X, Y, tuple(int(knd)...))
```
Comment (simonster) [inline base/fft/FFTW.jl]: Also deprecated

File: base/fft/dct.jl [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+for (pf, pfr, K, inplace) in ((:plan_dct, :plan_r2r, REDFT10, false),
+                              (:plan_dct!, :plan_r2r!, REDFT10, true),
+                              (:plan_idct, :plan_r2r, REDFT01, false),
+                              (:plan_idct!, :plan_r2r!, REDFT01, true))
+    @eval function $pf{T<:fftwNumber}(X::StridedArray{T}, region; kws...)
+        r = [1:n for n in size(X)]
+        nrm = sqrt(0.5^length(region) * normalization(X,region))
+        DCTPlan{T,$K,$inplace}($pfr(X, $K, region; kws...), r, nrm,
+                               ntuple(length(region), i -> int(region[i])))
```
Comment (simonster) [inline base/fft/dct.jl]: Also deprecated

**Comment by ScottPJones** [comment]
@stevengj The fifth item, about performance, should be checked off though (like the lined-out comment)?
I do hope this gets into 0.4, I do like the design, and the FFT stuff that it does have might be enough for us already without FFTW.

**Comment by stevengj** [comment]
@ScottPJones, the performance isn't crucial for merging as long as we still have FFTW for all of the common types, which is why I thought it didn't need to be checked off.

**Comment by simonster** [comment]
@stevengj Since #11486 was merged, I think you'll also have to switch the argument order of `ntuple` in this PR to avoid deprecation warnings.

**Comment by ScottPJones** [comment]
Finally back! @stevengj Sorry, I didn't make myself clear, I meant that it looked like the performance issue mentioned in the top comment had been taken care of, and that somebody had struck out the text about it, but had simply forgotten to check off the box... I didn't mean at all that this PR should be held up for any reason except for any remaining things like deprecation warnings.  It looks like a very worthwhile improvement to the code... I like that it uses a more flexible framework... and for people who *can't* use FFTW, the decent performance it has is still infinitely better than *NO* performance at all.

**Comment by stevengj** [comment]
Okay, changed the `ntuple` usage.

**Comment by jtravs** [comment]
I'd just like to register a vote here for merging this soon, even if performance and other issues need to be resolved later. I'm currently having to merge this pull request into a local julia repository and it is a bit of a pain TBH.

**Comment by simonster** [comment]
Yes, +1 for merging

File: base/utf8proc.jl (hunk 2 above)
Comment (tkelman) [inline base/utf8proc.jl]: would be cleaner for that change to at least be a separate commit

**Comment by jtravs** [comment]
As an additional comment, is there any chance of getting a nicer syntax for preallocated output than ```A_mul_B!(y, p, x)```? I really like the ```y = P*x``` syntax (and the style of this PR in general), but I really need fast, preallocated, real ffts in my code and ```A_mul_B!(y, p, x)``` isn't very clean or clear. I guess I could write a macro... I also realize that this is a general issue for inplace operations, not just the DFT.

**Comment by nalimilan** [comment]
@jtravs Yes, see https://github.com/JuliaLang/julia/issues/249.

**Comment by simonbyrne** [comment]
It should also be possible to make [InplaceOps.jl](https://github.com/simonbyrne/InplaceOps.jl) work for this.

**Comment by stevengj** [comment]
Closing this as the new DFT api has been merged separately, and now a separate patch with the pure-Julia FFTs is needed.

**Comment by yuyichao** [comment]
I've rebased this branch, should I push it out?

**Comment by stevengj** [comment]
Please do, thanks @yuyichao.

**Comment by yuyichao** [comment]
The rebase is [here](https://github.com/JuliaLang/julia/tree/yyc/dftnew_rebase). I simply rebased it and checked all the conflicts.

I didn't check the compatibility with the current master of the non-conflict parts and I also didn't squash any of the non-empty commits.

**Comment by hayd** [comment]
@yuyichao Bump! Is this going to be a PR (now that 0.4 is branched) ? :)

**Comment by yuyichao** [comment]
I don't think I'm familiar with this enough to do the PR. I rebased the branch to make sure the part I want to get into 0.4 doesn't mess up the rest of this PR too much.

CC. @stevengj

**Comment by stevengj** [comment]
It's not a priority for me at the moment, but it is certainly something that could be rebased onto 0.5 if needed.

**Comment by bjarthur** [comment]
sorry for reviving a now 4+ year old thread, but is there a differentiable FFT in julia?  IIUC, FFTW.jl is a C wrapper, so it is not, plus it is also GPL.  thanks.

**Comment by dhairyagandhi96** [comment]
We have the adjoints for FFTW.jl in Zygote.jl already, and will change that to AbstractFFT.jl very soon too

https://github.com/FluxML/Zygote.jl/blob/17ca911b82134c4a765822cd2b7ee19e959cc8e4/src/lib/array.jl#L777 for reference
